### PR TITLE
refactor TRA

### PR DESCRIPTION
**Pull request by evanzd**
Refactor TRA to support using Qlib dataset. 

## Motivation and Context
The current implementation cannot support Qlib dataset like `Alpha158` and `Alpha360`. In this PR, we have improved `MTSDatasetH` that is compatible to these datasets. Besides, we also support transport samples as daily groups and support `oracle transport` that discovers the optimal assignment for model diversity.

## How Has This Been Tested?
Has been tested by running the accompanied config files: `workflow_config_tra_Alpha360.yaml`, `workflow_config_tra_Alpha158.yaml` and `workflow_config_tra_Alpha158_full.yaml` .

## Screenshots of Test Results (if appropriate):
Running `workflow_config_tra_Alpha360.yaml` with `seed = 6000`:
image

## Types of changes
- [ ] Fix bugs
- [X] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: qlib/data/dataset/loader.py [hunk 1]
```diff
@@ -199,6 +199,10 @@ def __init__(self, config: dict, join="outer"):
         self.join = join
         self._data = None
 
+    def __getstate__(self) -> dict:
```
Comment (evanzd) [inline qlib/data/dataset/loader.py:202]: Add `__getstate__` to avoid saving `data_loader._data` to mlruns.

File: qlib/contrib/data/dataset.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+                for slc in slices_subset:
+
+                    # legacy support for Alpha360 data by `input_size`
+                    if self.input_size:
+                        data.append(self._data[slc.stop - 1].reshape(self.input_size, -1).T)
+                    else:
+                        data.append(_maybe_padding(self._data[slc], self.seq_len, self._zeros))
+
+                    if self.memory_mode == "sample":
+                        state.append(_maybe_padding(self._memory[slc][: -self.horizon], self.seq_len, self._zeros))
```
Comment (evanzd) [inline qlib/contrib/data/dataset.py:322]: move `self.horizon` out of `_maybe_padding`

File: qlib/contrib/model/pytorch_tra.py [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+        if "TPE" in self.src_info:
+            out = self.router(hist_loss)[0][:, -1]  # TPE
+            if "LR" in self.src_info:
+                out = torch.cat([hidden, out], dim=-1)  # LR_TPE
+        else:
+            out = hidden  # LR
+
+        out = self.fc(out)
+
+        choice = F.gumbel_softmax(out, dim=-1, tau=self.tau, hard=True)
```
Comment (evanzd) [inline qlib/contrib/model/pytorch_tra.py:662]: maybe `hard=False` is better

File: qlib/contrib/model/pytorch_tra.py [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+
+        evals_result["train"] = []
+        evals_result["valid"] = []
+        evals_result["test"] = []
+
+        epoch = 0
+        if self.pretrain:
+
+            self.logger.info("pretraining...")
+            self.optimizer = optim.Adam(list(self.model.parameters()) + list(self.tra.parameters()), lr=self.lr)
```
Comment (evanzd) [inline qlib/contrib/model/pytorch_tra.py:389]: pretrain only need `self.model.parameters()`

File: qlib/contrib/model/pytorch_tra.py [hunk 5]
```diff
@@ -0,0 +1,825 @@
...
+            _, epoch = self._fit(train_set, valid_set, test_set, evals_result, is_pretrain=True)
+
+            self.logger.info("reset TRA")
+            self.tra.reset_parameters()  # reset both router and predictors
+
```
Comment (evanzd) [inline qlib/contrib/model/pytorch_tra.py:394]: can remove this as tra is not optimized

File: qlib/contrib/model/pytorch_tra.py [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+            self.optimizer = optim.Adam(self.tra.parameters(), lr=self.lr)  # optimize TRA only
+
```
Comment (evanzd) [inline qlib/contrib/model/pytorch_tra.py:396]: should we add `self.model.parameters()`?

File: qlib/contrib/model/pytorch_tra.py [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+        epoch = 0
+        if self.pretrain:
+
+            self.logger.info("pretraining...")
+            self.optimizer = optim.Adam(list(self.model.parameters()) + list(self.tra.parameters()), lr=self.lr)
+            _, epoch = self._fit(train_set, valid_set, test_set, evals_result, is_pretrain=True)
+
+            self.logger.info("reset TRA")
+            self.tra.reset_parameters()  # reset both router and predictors
```
Comment (evanzd) [inline qlib/contrib/model/pytorch_tra.py:393]: shall we keep predictors?

Comment (linhx25) [reply qlib/contrib/model/pytorch_tra.py:393]: It seems that it makes no difference whether keep predictors
- self.tra.router.reset_parameters()
- the learning curve shows few differences.

File: qlib/contrib/model/pytorch_tra.py (hunk 3 above)
Comment (linhx25) [reply qlib/contrib/model/pytorch_tra.py:662]: It seems that changing `hard=False` makes no difference from the learning curve.

File: qlib/contrib/model/pytorch_tra.py (hunk 7 above)
Comment (linhx25) [reply qlib/contrib/model/pytorch_tra.py:393]: Whether keeping predictors' learned parameters makes no difference.
image

File: qlib/contrib/model/pytorch_tra.py (hunk 3 above)
Comment (evanzd) [reply qlib/contrib/model/pytorch_tra.py:662]: `hard=True` makes sure hard assignment

**Comment by evanzd** [comment]
@you-n-g Please help review this PR at your convenience. Thanks :)

**Comment by liufeirong** [comment]
In the  predict  function,it  called the  test_epoch method  which  used the  label col.In  the  truth, when  predict, the  label is None
